doc/developer/scripting.rst

```diff
@@ -269,57 +215,114 @@ has). For example, here is the encoder function for ``struct prefix``:
+   {
+        lua_getfield(L, idx, "network");
+        (void)str2prefix(lua_tostring(L, -1), prefix);
+        lua_pop(L, 1);
+        /* pop the table */
+        lua_pop(L, 1);
+   }
+
+.. warning::
+   ``lua_decode_prefix`` functions should leave the Lua stack completely empty
+   when they return.
+   For decoders that unmarshall fields from tables, remember to pop the table
+   at the end.
+
+
+A ``lua_to*`` function perform a similar role except that it first allocates
+memory for the new C type before decoding the value from the Lua stack, then
+returns a pointer to the newly allocated C type.
+This function can and should be implemented using ``lua_decode_*``:
 
 .. code-block:: c
 
    void *lua_toprefix(lua_State *L, int idx)
    {
            struct prefix *p = XCALLOC(MTYPE_TMP, sizeof(struct prefix));
 
-           lua_getfield(L, idx, "network");
-           str2prefix(lua_tostring(L, -1), p);
-           lua_pop(L, 1);
-
+           lua_decode_prefix(L, idx, p);
            return p;
    }
 
-By convention these functions should be called ``lua_to*``, as this is the
-naming convention used by the Lua C library for the basic types e.g.
-``lua_tointeger`` and ``lua_tostring``.
 
 The returned data must always be copied off the stack and the copy must be
 allocated with ``MTYPE_TMP``. This way it is possible to unload the script
 (destroy the state) without invalidating any references to values stored in it.
+Note that it is the caller's responsibility to free the data.
 
-To register a new type with its corresponding encoding functions:
+For consistency, we should always name functions of the first type
+``lua_decode_*``.
+Functions of the second type should be named ``lua_to*``, as this is the
+naming convention used by the Lua C library for the basic types e.g.
+``lua_tointeger`` and ``lua_tostring``.
+
+This two-function design allows the compiler to warn if a value passed into
+``frrscript_call`` does not have a encoder and decoder for that type.
+The ``lua_to*`` functions enable us to easily create decoders for nested
+structures.
+
+To register a new type with its corresponding encoding and decoding functions,
+add the mapping in the following macros in ``frrscript.h``:
 
 .. code-block:: c
 
-   struct frrscript_codec frrscript_codecs_lib[] = {    
-             {.typename = "prefix",    
-              .encoder = (encoder_func)lua_pushprefix,    
-              .decoder = lua_toprefix},    
-             {.typename = "sockunion",    
-              .encoder = (encoder_func)lua_pushsockunion,    
-              .decoder = lua_tosockunion},    
-              ...
-              {}};
+   #define ENCODE_ARGS_WITH_STATE(L, value) \
+          _Generic((value), \
+          ...
+   struct prefix * : lua_pushprefix, \
+   )(L, value)
+
+   #define DECODE_ARGS_WITH_STATE(L, value) \
+          _Generic((value), \
+          ...
+   struct prefix * : lua_decode_prefix, \
+   )(L, -1, value)
```

Comment:
I suggest formatting this as a patch (with +/- in the leading column) and use the `diff` markup specifier instead of `c`.

Reply:
updated

Reply:
This isn't quite right - the +/- should be in a column by themselves, like this:

```diff
  #define ENCODE_ARGS_WITH_STATE(L, value) \
         _Generic((value), \
         ...
- struct peer * : lua_pushpeer \
+ struct peer * : lua_pushpeer, \
+ struct prefix * : lua_pushprefix \
  )(L, value)

  #define DECODE_ARGS_WITH_STATE(L, value) \
         _Generic((value), \
         ...
- struct peer * : lua_decode_peer \
+ struct peer * : lua_decode_peer, \
+ struct prefix * : lua_decode_prefix \
  )(L, -1, value)
```

Reply:
Ah I see. updated